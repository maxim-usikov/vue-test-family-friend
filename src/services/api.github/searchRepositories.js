const {
  VUE_APP_GITHUB_API_ENDPOINT,
  VUE_APP_GITHUB_PERSONAL_ACCESS_TOKEN,
} = process.env;

function createQuery({ q, first = 10 }) {
  const query = `
query searchRepositories($q: String!, $first: Int!) {
  search(query: $q, type: REPOSITORY, first: $first) {
    edges {
      node {
        ... on Repository {
          id
          url
          name
          nameWithOwner
          description
          homepageUrl
          licenseInfo {
            name
          }
          stargazers {
            totalCount
          }
        }
      }
      cursor
    }
    pageInfo {
      endCursor
      hasNextPage
      hasPreviousPage
      startCursor
    }
    repositoryCount
  }
}
  `;

  return Object.assign({}, { query }, { variables: { q, first } });
}

async function searchRepositories(q) {
  const response = await fetch(VUE_APP_GITHUB_API_ENDPOINT, {
    method: 'POST',
    mode: 'cors',
    cache: 'no-cache',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `bearer ${VUE_APP_GITHUB_PERSONAL_ACCESS_TOKEN}`,
    },
    body: JSON.stringify(createQuery({ q })),
  });

  return response.json();
}

export default searchRepositories;
