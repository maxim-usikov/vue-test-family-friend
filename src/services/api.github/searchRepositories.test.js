import searchRepositories from './searchRepositories';

function mockFetch(data) {
  return jest.fn().mockImplementation(() => Promise.resolve({
    ok: true,
    json: () => data,
  }));
}

global.fetch = mockFetch({});

describe('searchRepositories', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('calls fetch', async () => {
    expect.assertions(2);

    await searchRepositories('test');

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toMatchSnapshot();
  });
});
