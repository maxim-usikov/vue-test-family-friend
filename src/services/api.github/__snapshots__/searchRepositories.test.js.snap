// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchRepositories calls fetch 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.github.com/graphql",
      Object {
        "body": "{\\"query\\":\\"\\\\nquery searchRepositories($q: String!, $first: Int!) {\\\\n  search(query: $q, type: REPOSITORY, first: $first) {\\\\n    edges {\\\\n      node {\\\\n        ... on Repository {\\\\n          id\\\\n          url\\\\n          name\\\\n          nameWithOwner\\\\n          description\\\\n          homepageUrl\\\\n          licenseInfo {\\\\n            name\\\\n          }\\\\n          stargazers {\\\\n            totalCount\\\\n          }\\\\n        }\\\\n      }\\\\n      cursor\\\\n    }\\\\n    pageInfo {\\\\n      endCursor\\\\n      hasNextPage\\\\n      hasPreviousPage\\\\n      startCursor\\\\n    }\\\\n    repositoryCount\\\\n  }\\\\n}\\\\n  \\",\\"variables\\":{\\"q\\":\\"test\\",\\"first\\":10}}",
        "cache": "no-cache",
        "headers": Object {
          "Authorization": "bearer test",
          "Content-Type": "application/json",
        },
        "method": "POST",
        "mode": "cors",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
